README.md

```diff
@@ -1,12 +1,332 @@
+
+
+### With LinearOperator
+
+The LinearOperator package offers the best of both worlds:
+
+```python
+from linear_operator.operators import DiagLinearOperator, RootLinearOperator
+# C = torch.randn(1000, 20)
+# d = torch.randn(1000)
+# b = torch.randn(1000)
+A = RootLinearOperator(C) + DiagLinearOperator(d)  # represents C C^T + diag(d)
+```
+
+it provides an interface that lets us treat $\boldsymbol A$ as if it were a generic tensor,
+using the standard PyTorch API...
+
+```python
+torch.linalg.solve(A, d, b)  # computes A^{-1} b efficiently!
 ```
-python -m unittest discover
+
+but under-the-hood, the `LinearOperator` object keeps track of the algebraic structure of $\boldsymbol A$ (low rank plus diagonal)
+and determines the most efficient routine to use (the Woodbury formula).
+This way, we can get a efficient ($\mathcal O(N)$) solve while abstracting away all of the details.
+
+Crucially, $\boldsymbol A$ is never explicitly instantiated as a matrix, which makes it possible to scale
+to very large operators without running out of memory:
+
+```python
+C = torch.randn(10000000, 20)
+d = torch.randn(10000000)
+b = torch.randn(10000000)
+A = RootLinearOperator(C) + DiagLinearOperator(d)  # represents a 10M x 10M matrix!
+torch.linalg.solve(A, d, b)  # computes A^{-1} b efficiently!
 ```
+
+### What is a Linear Operator?
+A linear operator is a generalization of a matrix.
+It is a linear function that is defined in by its application to a vector.
+The most common linear operators are (potentially structured) matrices,
+where the function applying them to a vector are (potentially efficient)
+matrix-vector multiplication routines.
+
+In code, a `LinearOperator` is a class that
+1. specifies the tensor(s) needed to define the LinearOperator,
+1. specifies a `_matmul` function (how the LinearOperator is applied to a vector),
+1. specifies a `_size` function (how big is the LinearOperator if it is represented as a matrix, or batch of matrices), and
+1. specifies a `_transpose_nonbatch` function (the adjoint of the LinearOperator).
+
+For example:
+
+```python
+class DiagLinearOperator(linear_operator.LinearOperator):
+    r"""
+    A LinearOperator representing a diagonal matrix.
+    """
+    def __init__(self, diag):
+        # diag: the vector that defines the diagonal of the matrix
+        self.diag = diag
+
+    def _matmul(self):
+        return torch.Size([*self.diag.shape, self.diag.size(-1)])
+
+    def _size(self):
+        return torch.Size([*self.diag.shape, self.diag.size(-1)])
+
+    def _transpose_nonbatch(self):
+        return self  # Diagonal matrices are symmetric
+
+
+# ...
+
+D = DiagLinearOperator(torch.tensor([1., 2., 3.])
+# Represents the matrix
+#   [[1., 0., 0.],
+#    [0., 2., 0.],
+#    [0., 0., 3.]]
+torch.matmul(D, torch.tensor([4., 5., 6.])
+# Returns [4., 10., 18.]
+```
+
+Other functions (e.g. `logdet`, `eigh`, etc.) can be defined as needed if efficient routines exist.
```

Comment:
We should make this more prominent, this is easily skipped in the current layout in which case the package seems a lot less impressive / useful.

Reply:
Suggestions for where to put it?

Reply:
you could add another list item above so it's not as split: 

1. specifies a `_transpose_nonbatch` function (the adjoint of the LinearOperator).
1. (optionally) defines other functions (e.g. `logdet`, `eigh`, etc.) to accelerate computations for which efficient structure-exploiting routines exist.